File: src/services/openai/completions.ts
import OpenAI from 'openai'
import { OPENAI_API_KEY, IS_OPENAI_ENABLED } from '@/config/constants'
import type { AnimeRecommendation } from '@/types/openai'

let openai: OpenAI | null = null

if (IS_OPENAI_ENABLED && OPENAI_API_KEY) {
  openai = new OpenAI({
    apiKey: OPENAI_API_KEY
  })
}

/**
 * Find anime recommendations based on a description using ChatGPT-5
 */
export async function findAnimeByDescription(description: string): Promise<AnimeRecommendation[]> {
  if (!openai || !IS_OPENAI_ENABLED) {
    throw new Error('OpenAI is not configured. Please set OPENAI_API_KEY environment variable.')
  }

  try {
    const prompt = `Find 3 anime that match: "${description}"

Respond with only valid JSON:
[
  {"title": "Anime Name", "reason": "Why it matches", "confidence": 0.9}
]`

    const completion = await openai.chat.completions.create({
      model: 'gpt-5',
      messages: [
        {
          role: 'user',
          content: prompt
        }
      ]
      //   max_completion_tokens: 1000
    })

    // console.log('OpenAI completion:', JSON.stringify(completion, null, 2))

    const response = completion.choices[0]?.message?.content
    if (!response) {
      console.error('No response content. Full completion:', completion)
      throw new Error('No response from OpenAI')
    }

    // console.log('OpenAI response:', response)

    // Parse the JSON response
    const recommendations = JSON.parse(response) as AnimeRecommendation[]

    // Validate the response structure
    if (!Array.isArray(recommendations)) {
      throw new Error('Invalid response format')
    }

    return recommendations.filter(
      (rec) => rec.title && rec.reason && typeof rec.confidence === 'number'
    )
  } catch (error) {
    console.error('Error finding anime by description:', error)
    throw new Error('Failed to find anime recommendations')
  }
}

File: go/internal/services/openai/completions.go
package openai

import (
	"context"
	"encoding/json"
	"fmt"

	"discord-anime-bot/internal/types"

	"github.com/openai/openai-go/v2"
	"github.com/openai/openai-go/v2/option"
)

// FindAnimeByDescription uses OpenAI to find anime recommendations based on description
func FindAnimeByDescription(description, apiKey string) ([]types.OpenAIRecommendation, error) {
	if apiKey == "" {
		return nil, fmt.Errorf("OpenAI is not configured. Please set OPENAI_API_KEY environment variable")
	}

	client := openai.NewClient(option.WithAPIKey(apiKey))

	prompt := fmt.Sprintf(`Based on this description: "%s"

Please recommend anime titles that match this description. Return your response as a JSON array of objects with the following structure:
[
  {
    "title": "Exact anime title",
    "reason": "Brief explanation of why this matches",
    "confidence": 0.95
  }
]

Guidelines:
- Return 1-3 recommendations
- Use exact anime titles (romaji or English)
- Confidence should be between 0.0 and 1.0
- Only return valid JSON, no other text
- Focus on popular/well-known anime`, description)

	resp, err := client.Chat.Completions.New(
		context.Background(),
		openai.ChatCompletionNewParams{
			Model: openai.ChatModelGPT5, // Uses the latest model
			Messages: []openai.ChatCompletionMessageParamUnion{
				openai.UserMessage(prompt),
			},
		},
	)

	if err != nil {
		return nil, fmt.Errorf("OpenAI API error: %w", err)
	}

	if len(resp.Choices) == 0 {
		return nil, fmt.Errorf("no response from OpenAI")
	}

	content := resp.Choices[0].Message.Content

	var recommendations []types.OpenAIRecommendation
	if err := json.Unmarshal([]byte(content), &recommendations); err != nil {
		return nil, fmt.Errorf("failed to parse OpenAI response: %w", err)
	}

	return recommendations, nil
}
